CMakeLists.txt

```diff
@@ -10,6 +10,8 @@ include("cmake/tests.cmake")
 include("cmake/ccache.cmake")
 include("cmake/codesigning.cmake")
 
+set(CMAKE_CXX_STANDARD 17)
```

Comment:
Setting CMake variables directly is deprecated, target-based properties should be used instead:
```cmake
target_compile_features(target_name <PUBLIC|PRIVATE|INTERFACE> cxx_std_17)
```

Reply:
> Setting CMake variables directly is deprecated

Since when? Do you have sources for this? The official documentation for [`CMAKE_CXX_STANDARD`](https://cmake.org/cmake/help/latest/variable/CMAKE_CXX_STANDARD.html) does not mention a deprecation, neither does the entry on the [`CXX_STANDARD`](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html) property. There's also no hint in the documentation of [compile features](https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#optional-compile-features) that `CMAKE_CXX_STANDARD` is somehow discouraged from use.

I think removing a way of defining a default for populating target properties would maintaining large projects cumbersome and error prone, because you would have to repeat yourself over and over again instead of setting the required language version once.

To come back to why I'm setting this property in the main CMake file (as opposed to setting it in the Broker subdirectory CMake file): the language version in use may affect the ABI of the standard library. With C++11, GCC had to break its ABI for `std::string` in order to comply to the new small string optimization. While the committee is generally trying to avoid breaking the ABI, there's also a push for [establishing a policy for allowing ABI-incompatible changes](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1863r1.pdf). I deem it wise to make sure not to compile and link libraries that use different language versions.

On a more practical note, Broker explicitly [checks for C++17 and sets its flag accordingly](https://github.com/zeek/broker/blob/2740fd51dbd3bd57ae55ef6ae228d0f0709cb351/CMakeLists.txt#L101) if the parent did not provide suitable C++ flags, while CAF fully relies on the parent project to provide suitable C++ flags. Neither project uses `target_compile_features` to set a language version. And I think the Zeek agent should not first run the subdirectories and then modify targets of its subdirectory. That's only asking for trouble down the road and isn't maintainable, since targets may change at any point in time. The only targets the agent should rely on are the final library targets it needs for linking.

I think the best solution from a maintainer perspective is setting the language version *once* globally. We can't predict how the committee is going to decide on its ABI policy. C++23 may be incompatible with C++20/17 for all we know today. But we *do* know that compiling and linking everything with the same language version is always going to work.

Reply:
It is deprecated in modern CMake the same way some patterns are discouraged in modern C++.

The problem relies on the fact that global properties propagates and have effect on the whole project. A 'good citizen' library doesn't want that, because a project (such as osquery) may have its own settings and propagation logic. Additionally, the way it propagates is based on the scope, so it may not always be 100% clear which targets get it and which targets do not.

It is also discouraged to manually check for compiler features (whenever possible), because this is already supported and built into CMake using [per-target compiler requirements](https://cmake.org/cmake/help/latest/command/target_compile_features.html).

In general, setting things globally is a bad idea, especially for libraries that may be imported inside other libraries/programs.

Reply:
Repetition can be removed using the same pattern already in use by osquery and the Zeek Agent itself:

```cmake
add_library(thirdparty_broker_settings INTERFACE)
target_compile_features(thirdparty_broker_settings INTERFACE
  cxx_std_17
)

target_compile_options(thirdparty_broker_settings INTERFACE
  xxx
)
```
Linking the library against `thirdparty_broker_settings` will now automatically propagate the required settings from target to target based on visibility. This is more correct as it is not affected by scope, easier to understand/debug, and does not cause side-effects/issues when changing inclusion order

Reply:
I'd like to get this merged. @alessandrogario can you just go ahead and make the changes you suggest if you believe that's the better approach? That seems easier than further iterating here, as you already have the pattern to use in your head.

Reply:
> I'd like to get this merged. @alessandrogario can you just go ahead and make the changes you suggest if you believe that's the better approach? That seems easier than further iterating here, as you already have the pattern to use in your head.

Working on it!